response.go

```diff
@@ -0,0 +1,91 @@
+package jsonrpc
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+type response struct {
+	Jsonrpc string        `json:"jsonrpc"`
+	Result  interface{}   `json:"result,omitempty"`
+	ID      interface{}   `json:"id"`
+	Error   *JSONRPCError `json:"error,omitempty"`
+}
+
+func (r response) MarshalJSON() ([]byte, error) {
+	// Custom marshal logic as per JSON-RPC 2.0 spec:
+	// > `result`:
+	// > This member is REQUIRED on success.
+	// > This member MUST NOT exist if there was an error invoking the method.
+	//
+	// > `error`:
+	// > This member is REQUIRED on error.
+	// > This member MUST NOT exist if there was no error triggered during invocation.
+	data := map[string]interface{}{
+		"jsonrpc": r.Jsonrpc,
+		"id":      r.ID,
+	}
+
+	if r.Error != nil {
+		data["error"] = r.Error
+	} else {
+		data["result"] = r.Result
+	}
+	return json.Marshal(data)
+}
+
+type JSONRPCError struct {
+	Code    ErrorCode       `json:"code"`
+	Message string          `json:"message"`
+	Meta    json.RawMessage `json:"meta,omitempty"`
+	Data    interface{}     `json:"data,omitempty"`
+}
+
+func (e *JSONRPCError) Error() string {
+	if e.Code >= -32768 && e.Code <= -32000 {
+		return fmt.Sprintf("RPC error (%d): %s", e.Code, e.Message)
+	}
+	return e.Message
+}
+
+var (
+	_             error = (*JSONRPCError)(nil)
+	marshalableRT       = reflect.TypeOf(new(marshalable)).Elem()
+	errorCodecRT        = reflect.TypeOf(new(RPCErrorCodec)).Elem()
+)
+
+func (e *JSONRPCError) val(errors *Errors) reflect.Value {
+	if errors != nil {
+		t, ok := errors.byCode[e.Code]
+		if ok {
+			var v reflect.Value
+			if t.Kind() == reflect.Ptr {
+				v = reflect.New(t.Elem())
+			} else {
+				v = reflect.New(t)
+			}
+
+			if v.Type().Implements(errorCodecRT) {
+				if err := v.Interface().(RPCErrorCodec).FromJSONRPCError(*e); err != nil {
+					log.Errorf("Error converting JSONRPCError (code %d) to custom error type '%s': %w", e.Code, t.String(), err)
+					return reflect.ValueOf(e)
+				}
+			}
+
+			if len(e.Meta) > 0 && v.Type().Implements(marshalableRT) {
```

Comment:
```suggestion
			} else if len(e.Meta) > 0 && v.Type().Implements(marshalableRT) {
```

let's disallow dual behaviour for now, we could enable this at some point if there was ever a case for it (I can't think what it would be) but best avoid surprises if we're not testing for it

Reply:
Done.